server/http/server.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package http
+
+import (
+	"net/http"
+	"sync"
+
+	mserver "github.com/m3db/m3ctl/server"
+	"github.com/m3db/m3ctl/service"
+	"github.com/m3db/m3x/log"
+
+	"github.com/gorilla/mux"
+)
+
+const (
+	publicPathPrefix = "/public"
+	publicDirPath    = "public"
+	staticPathPrefix = "/static"
+	staticDirPath    = "ui/build/static"
+	indexFilePath    = "ui/build/index.html"
+)
+
+type server struct {
+	server   *http.Server
+	services []service.Service
+	logger   log.Logger
+	wg       sync.WaitGroup
+}
+
+// NewServer creates a new HTTP server.
+func NewServer(address string, opts Options, services ...service.Service) mserver.Server {
+	cloned := make([]service.Service, len(services))
```

Comment:
Why do we need to make a copy here?

Reply:
This is to prevent the list of services from being mutated externally after the server is constructed, e.g., see https://play.golang.org/p/YIZLwWGkmv0

Reply:
Ha, that is weird. Maybe add a comment just to make it clear?

Reply:
Yeah this is essentially the compiler trying to avoid extra allocation with variadic functions but yeah could potentially cause surprises, I'll add a comment.